File: cpp/src/include/classes/game_elements/cell/cell.cpp
#include "cell.h"

Cell::Cell(int init_x_coord, int init_y_coord) :
    x_coord_(init_x_coord),
    y_coord_(init_y_coord),
    is_hit_(false),
    is_occupied_(false),
    occupied_probability_(0)
{
}

int Cell::getXCoord() const {
    return x_coord_;
}

int Cell::getYCoord() const {
    return y_coord_;
}

bool Cell::getIsHit() const {
    return is_hit_;
}

bool Cell::getIsOccupied() const {
    return is_occupied_;
}

int Cell::getOccupiedProbability() const {
    return occupied_probability_;
}

void Cell::setIsHit() {
    if (is_hit_) {
        throw std::runtime_error("Cella già colpita");
    }
    is_hit_ = true;
}

void Cell::setIsOccupied() {
    if (is_occupied_) {
        throw std::runtime_error("Cella già occupata");
    }
    is_occupied_ = true;
}

void Cell::increaseOccupiedProbability(int n) {
    occupied_probability_ += n;
}

void Cell::resetOccupiedProbability() {
    occupied_probability_ = 0;
}

File: go/src/include/game_elements/cell.go
// Pacchetto game_elements fornisce elementi di gioco per il server di battaglia navale.
package game_elements

import (
	"errors"
)

// Cell rappresenta una singola cella.
type Cell struct {
	IsHit      bool `json:"is_hit"`
	IsOccupied bool `json:"is_occupied"`
}

// NewCell restituisce una nuova istanza di Cell inizializzata con i valori predefiniti.
func NewCell() *Cell {
	return &Cell{IsHit: false, IsOccupied: false}
}

// SetIsHit imposta la cella come colpita e restituisce true se la cella era occupata.
// Restituisce un errore se la cella è già stata colpita in precedenza.
func (c *Cell) SetIsHit() (bool, error) {
	if c.GetIsHit() {
		return false, errors.New("cella già colpita")
	}
	c.IsHit = true
	if c.IsOccupied {
		return true, nil
	}
	return false, nil
}

// SetIsOccupied imposta la cella come occupata e restituisce un errore se la cella è già occupata.
func (c *Cell) SetIsOccupied() error {
	if c.GetIsOccupied() {
		return errors.New("cella già occupata")
	}
	c.IsOccupied = true
	return nil
}

// GetIsHit restituisce true se la cella è stata colpita, altrimenti restituisce false.
func (c *Cell) GetIsHit() bool {
	return c.IsHit
}

// GetIsOccupied restituisce true se la cella è occupata, altrimenti restituisce false.
func (c *Cell) GetIsOccupied() bool {
	return c.IsOccupied
}
